clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -0,0 +1,591 @@
+                List<TopicPartition> toRemove = new ArrayList<>();
+                for (TopicPartition partition: entry.getValue())
+                    if (!partition2AllPotentialConsumers.keySet().contains(partition))
+                        toRemove.add(partition);
+                for (TopicPartition partition: toRemove)
+                    entry.getValue().remove(partition);
+            }
+            TreeSet<String> sortedConsumers = new TreeSet<>(new SubscriptionComparator(assignments));
+            sortedConsumers.addAll(assignments.keySet());
+
+            while (!sortedConsumers.isEmpty()) {
+                String consumer = sortedConsumers.pollLast();
+                List<TopicPartition> remainingPartitions = assignments.get(consumer);
+                if (!remainingPartitions.isEmpty()) {
+                    sortedPartitions.add(remainingPartitions.remove(0));
+                    sortedConsumers.add(consumer);
+                }
+            }
+
+            for (TopicPartition partition: partition2AllPotentialConsumers.keySet()) {
+                if (!sortedPartitions.contains(partition))
+                    sortedPartitions.add(partition);
+            }
+
+        } else {
+            // an ascending sorted set of topic partitions based on how many consumers can potentially use them
+            TreeSet<TopicPartition> sortedAllPartitions = new TreeSet<>(new PartitionComparator(partition2AllPotentialConsumers));
+            sortedAllPartitions.addAll(partition2AllPotentialConsumers.keySet());
+
+            while (!sortedAllPartitions.isEmpty())
+                sortedPartitions.add(sortedAllPartitions.pollFirst());
+        }
+
+        return sortedPartitions;
+    }
+
+    /**
+     * @param partition2AllPotentialConsumers a mapping of partitions to their potential consumers
+     * @param consumer2AllPotentialPartitions a mapping of consumers to potential partitions they can consumer from
+     * @return true if potential consumers of partitions are the same, and potential partitions consumers can
+     * consumer from are the same too
+     */
+    private boolean areSubscriptionsIdentical(HashMap<TopicPartition, List<String>> partition2AllPotentialConsumers,
+                                              HashMap<String, List<TopicPartition>> consumer2AllPotentialPartitions) {
+        if (!hasIdenticalListElements(partition2AllPotentialConsumers.values()))
+            return false;
+
+        if (!hasIdenticalListElements(consumer2AllPotentialPartitions.values()))
+            return false;
+
+        return true;
+    }
+
+    /**
+     * @param col a collection of elements that are lists themselves
+     * @return true if all lists in the collection have the same members; false otherwise
+     */
+    private <T> boolean hasIdenticalListElements(Collection<List<T>> col) {
+        Set<Set<T>> set = new HashSet<Set<T>>(col.size());
```

Comment:
Couldn't we could just iterate through the collection and ensure that each list equals the previous one.